### PR TITLE
MASP nullifier uniqueness

### PR DESCRIPTION
**Pull request by grarco**
## Describe your changes

Addresses #1373.

Implements a nullifier set to prevent double spending in masp transactions. Updates the masp VP to check that the transactions writes the correct nullifiers and that these haven't already been revealed.

**NOTE**: testing this with an integration or e2e test is impossible cause we cannot force the submission of the tx if the builder fails to construct the shielded `Transaction`. We'd need to test this with a unit test (which I think would be the correct way regardless of the `force` issue) but at the moment we don't have the tools to write unit tests for the masp vp (at least to my knowledge)

## Indicate on which release or other PRs this topic is based on

v0.27.0

## Checklist before merging to `draft`
- [x] I have added a changelog
- [x] Git history is in acceptable state

### CODE REVIEW

File: shared/src/ledger/native_vp/masp.rs [hunk 1]
```diff
@@ -161,6 +166,43 @@ where
...
+                    || revealed_nullifiers.contains(&nullifier_key)
+                {
+                    tracing::debug!(
+                        "MASP double spending attempt, the nullifier {:#?} \
+                         has already been revealed previously",
+                        description.nullifier.0
+                    );
+                    return Ok(false);
+                }
+                revealed_nullifiers.insert(nullifier_key);
```
Comment (murisi) [inline shared/src/ledger/native_vp/masp.rs:192]: Right now in the VP we check that the nullifier keys changed are a subset of those in the transaction. But where do we check that there is a key change for each of the nullifiers in the transaction? I would expect this check to be somewhere in this loop...

File: shared/src/ledger/native_vp/ibc/context.rs [hunk 2]
```diff
@@ -237,6 +239,19 @@ where
         );
         self.write(&current_tx_key, record.serialize_to_vec())?;
         self.write(&head_tx_key, (current_tx_idx + 1).serialize_to_vec())?;
+        for description in shielded
```
Comment (murisi) [inline shared/src/ledger/native_vp/ibc/context.rs:242]: I see a bit of duplication in how we update the nullifier set in the transaction codes. But maybe getting rid of it is easier said than done and can be deferred to a later PR?

**Review by murisi — CHANGES_REQUESTED** [review]
Do you know why the unit and integration tests are failing in the CI? They seem to be timing out for some reason... Do they pass on your machine? Have you tried regenerating the MASP test fixtures maybe?

File: shared/src/ledger/native_vp/ibc/context.rs (hunk 2 above)
Comment (grarco) [reply shared/src/ledger/native_vp/ibc/context.rs]: I've tried to improve the thing with a shared function now

File: shared/src/ledger/native_vp/masp.rs (hunk 1 above)
Comment (grarco) [reply shared/src/ledger/native_vp/masp.rs:202]: Nice catch thanks! I've added the missing check now

**Comment by grarco** [comment]
> Do you know why the unit and integration tests are failing in the CI? They seem to be timing out for some reason... Do they pass on your machine? Have you tried regenerating the MASP test fixtures maybe?

Both unit and integration tests pass on my machine without the need to rebuild the proofs

**Comment by grarco** [comment]
> Nice. Could you please just check out the `serialize_to_vec` issue in the comments to see if there is anything to it? And if you get the chance (but it's not strictly necessary), could you please try seeing if moving `handle_masp_tx` from `tx_prelude/src/token.rs` into `core/src/ledger/masp_utils.rs` is trivial or not?

Ok' I've moved the entire `handle_masp_tx` to the `masp_utils` file. This has also solved the double serialization issue (which yes, I believe like you do it was a bug). To be able to share this function also with the IBC context I had to remove the call to `insert_verifier(masp_addr)` from the function (others I would have needed to implement `TxEnv` on the IBC context), which shouldn't be a problem because the tx writes some masp keys anyway so those should already trigger the vp

**Pull request merged**